### ISSUE TITLE
Should the `ready` future fail if the socket gets closed while not resolved?

### ISSUE THREAD

**Issue by davidmartos96** [dart-lang/http#1693]
As the title suggests, would it make sense to throw when stablishing the connection and in the meantime some piece of code closes the socket?

On Node that is precisely what happens in the `ws` package.

Kind of related to this use case, it would be helpful to have a getter on the channel that says if it's closed or not.

Here is a simple demo for the close while connecting case on Dart and NodeJS:

```dart
import 'dart:async';

import 'package:web_socket_channel/web_socket_channel.dart';

Future<void> main() async {
  final wsUrl = Uri.parse('ws://localhost:9898');
  final channel = WebSocketChannel.connect(wsUrl);

  final readyFuture = channel.ready;

  unawaited(Future(() => channel.sink.close()));

  await readyFuture;

  channel.stream.listen((message) {
    channel.sink.add('received!');
  });
}
```

```js
import { WebSocket } from "ws";

const sleepAsync = (ms: number) =>
  new Promise((resolve) => setTimeout(resolve, ms));

async function main() {
  const socket = new WebSocket("ws://localhost:9898");
  socket.binaryType = "nodebuffer";

  socket.addEventListener("open", () => {
    console.log("Connected to server");
  });
  socket.addEventListener("message", (m) => {
    console.log("Message:", m);
  });
  socket.addEventListener("error", (e) => {
    console.log("Error obtained:", e.message);
  });
  socket.addEventListener("close", () => {
    console.log("Closed");
  });
  socket.close();
  await sleepAsync(1000);
}

main();
```

Output on Node:
```
Error obtained: WebSocket was closed before the connection was established
Closed
```

**Comment by kashyap-5paisa** [comment]
Yep, I’m facing the exact same issue! If a WebSocket is closed while still connecting, Dart just shrugs, whereas in Android (Kotlin) and Node.js (ws package), it throws an error.

To make things worse, when the connection doesn’t exist, await channel.sink.close() sometimes just… never responds. I had to hardcode a timeout to prevent my app from freezing:

`await channel?.sink.close(WebStatus.normalClosure).timeout(const Duration(milliseconds: 300));` 

Your demo captures the issue perfectly. Dart should:

- Provide a reliable way to check if the WebSocket is open.
- Ensure sink.close() doesn’t hang indefinitely.

Right now, handling WebSockets in Dart feels like trying to keep a conversation going with someone who’s already left the room. 

### In short:

As @davidmartos96 rightly said, simple isClosed getter within websocket channel will be a lifesaver

@flutter team please check this
